compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala

```diff
@@ -0,0 +1,336 @@
+
+          // f(args)
+          case tree: Apply =>
+            if canInstrumentApply(tree) then
+              if needsLift(tree) then
+                instrumentLifted(tree)
+              else
+                instrument(transformApply(tree))
+            else
+              transformApply(tree)
+
+          // (f(x))[args]
+          case TypeApply(fun: Apply, args) =>
+            cpy.TypeApply(tree)(transform(fun), args)
+
+          // a.b
+          case Select(qual, name) =>
+            if qual.symbol.exists && qual.symbol.is(JavaDefined) then
+              //Java class can't be used as a value, we can't instrument the
+              //qualifier ({<Probe>;System}.xyz() is not possible !) instrument it
+              //as it is
+              instrument(tree)
+            else
+              val transformed = cpy.Select(tree)(transform(qual), name)
+              if transformed.qualifier.isDef then
+                // instrument calls to methods without parameter list
+                instrument(transformed)
+              else
+                transformed
+
+          case tree: CaseDef => instrumentCaseDef(tree)
+          case tree: ValDef =>
+            // only transform the rhs
+            val rhs = transform(tree.rhs)
+            cpy.ValDef(tree)(rhs = rhs)
+
+          case tree: DefDef =>
+            // Only transform the params (for the default values) and the rhs.
+            val paramss = transformParamss(tree.paramss)
+            val rhs = transform(tree.rhs)
+            val finalRhs =
+              if canInstrumentDefDef(tree) then
+                // Ensure that the rhs is always instrumented, if possible
+                instrumentBody(tree, rhs)
+              else
+                rhs
+            cpy.DefDef(tree)(tree.name, paramss, tree.tpt, finalRhs)
+
+          case tree: PackageDef =>
+            // only transform the statements of the package
+            cpy.PackageDef(tree)(tree.pid, transform(tree.stats))
+          case tree: Assign =>
+            // only transform the rhs
+            cpy.Assign(tree)(tree.lhs, transform(tree.rhs))
+
+          // For everything else just recurse and transform
+          // Special care for Templates: it's important to set the owner of the `stats`, like super.transform
+          case _ =>
+            super.transform(tree)
+        }
+
+    /** Lifts and instruments an application.
+      * Note that if only one arg needs to be lifted, we just lift everything.
+      */
+    private def instrumentLifted(tree: Apply)(using Context) =
+      // lifting
+      val buffer = mutable.ListBuffer[Tree]()
+      val liftedApply = LiftCoverage.liftForCoverage(buffer, tree)
+
+      // instrumentation
+      val instrumentedArgs = buffer.toList.map(transform)
+      val instrumentedApply = instrument(liftedApply)
+      Block(
+        instrumentedArgs,
+        instrumentedApply
+      ).withSpan(instrumentedApply.span)
```

Comment:
By default the span of a tree is the union of the spans of its children, so I don't think withSpan is needed here.

Reply:
good to know :)